wannierberri/__system.py

```diff
@@ -284,23 +289,130 @@ def set_symmetry(self,symmetry_gen=[]):
+                            )
+                        for (s,a,b) in [(+1,alpha_A,beta_A) , (-1,beta_A,alpha_A)] )
+                norm = np.linalg.norm(CC_R_new - self.conj_XX_R(CC_R_new))
+                assert norm<1e-10 , f"norm={norm}"
+
+
+            # not sure if the following is correct (Stepan)
+            if hasattr(self,'SA_R'):
+                assert hasattr(self,'SS_R') , "if you use SA_R, you need also SS_R"
+                SA_R_new  =  self.SA_R.copy() - self.SS_R[:,:,:,:,None]*self.wannier_centers_cart[None,:,None,None,:]
+            if hasattr(self,'SHA_R'):
+                assert hasattr(self,'SH_R') , "if you use SA_R, you need also SH_R"
+                SHA_R_new  =  self.SHA_R.copy() - self.SS_R[:,:,:,:,None]*self.wannier_centers_cart[None,:,None,None,:]
+
+            for X in ['AA','BB','CC','SA','SHA']:
+                if hasattr(self,X+'_R'):
+                    vars(self)[X+'_R'] = locals()[X+'_R_new']
+
+            for X in ['SA','SHA','SR','SH','SHR']:
+                if hasattr(self,X+'_R'):
+                    pass
+#                    raise NotImplementedError(f"use_wcc_phases=True is not implemented for {X}_R")
+
+
+    @property
+    def iR0(self):
+        return self.iRvec.tolist().index([0,0,0])
+
+    @lazy_property.LazyProperty
+    def reverseR(self):
+        """maps the R vector -R"""
+        iRveclst= self.iRvec.tolist()
+        mapping = np.all( self.iRvec[:,None,:]+self.iRvec[None,:,:] == 0 , axis = 2 )
+        # check if some R-vectors do not have partners
+        notfound = np.where(np.logical_not(mapping.any(axis=1)))[0]
+        for ir in notfound:
+            print ("WARNING : R[{}] = {} does not have a -R partner".format(ir,self.iRvec[ir]) )
+        # check if some R-vectors have more then 1 partner 
+        morefound = np.where(np.sum(mapping,axis=1)>1)[0]
+        if len(morefound>0):
+            raise RuntimeError( "R vectors number {} have more then one negative partner : \n{} \n{}".format(
+                            morefound,self.iRvec[morefound],np.sum(mapping,axis=1) ) )
+        lst1,lst2=[],[]
+        for ir1 in range(self.nRvec):
+            ir2 = np.where(mapping[ir1])[0]
+            if len(ir2)==1:
+                lst1.append(ir1)
+                lst2.append(ir2[0])
+#                print (ir1,self.iRvec[ir1] , ir2,self.iRvec[ir2[0]])
+        return np.array(lst1),np.array(lst2)
+
+    def conj_XX_R(self,XX_R):
+        """ reverses the R-vector and takes the hermitian conjugate """
+        XX_R_new = np.zeros_like(XX_R)
+        lst1,lst2 = self.reverseR
+        assert np.all(self.iRvec[lst1] + self.iRvec[lst2] ==0 )
+#        print (XX_R.shape,XX_R_new.shape,lst1,lst2)
+        XX_R_new [:,:,lst1] = np.copy(XX_R)[:,:,lst2]
+        XX_R_new[:] = XX_R_new.swapaxes(0,1).conj()
+        return np.copy(XX_R_new)
```

Comment:
Why return a copy?

Reply:
If contiguity is a problem, you can use https://numpy.org/doc/stable/reference/generated/numpy.ascontiguousarray.html